File: android/wechat/src/main/java/software/sitb/react/wechat/WeChatManager.java
package software.sitb.react.wechat;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;
import com.facebook.react.bridge.*;
import com.facebook.react.common.MapBuilder;
import com.tencent.mm.opensdk.modelbase.BaseReq;
import com.tencent.mm.opensdk.modelbase.BaseResp;
import com.tencent.mm.opensdk.modelmsg.*;
import com.tencent.mm.opensdk.modelpay.PayResp;
import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;
import software.sitb.react.commons.DefaultReactContextBaseJavaModule;

import javax.annotation.Nullable;
import java.io.IOException;
import java.net.URL;
import java.util.Map;
import java.util.UUID;

/**
 * @author Sean sean.snow@live.com createAt 2017/5/23
 */
public class WeChatManager extends DefaultReactContextBaseJavaModule implements IWXAPIEventHandler {

  private static final String TAG = "WeChatManager";

  private IWXAPI weChatApi;

  public WeChatManager(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  public String getName() {
    return "WeChatManager";
  }

  @Nullable
  @Override
  public Map<String, Object> getConstants() {
    return MapBuilder.<String, Object>of(
      "type", MapBuilder.of(
        "text", ShareType.TEXT.ordinal(),
        "image", ShareType.IMAGE.ordinal(),
        "music", ShareType.MUSIC.ordinal(),
        "video", ShareType.VIDEO.ordinal(),
        "webPage", ShareType.WEBPAGE.ordinal()
      ),
      "scene", MapBuilder.of(
        "session", SendMessageToWX.Req.WXSceneSession,
        "timeLine", SendMessageToWX.Req.WXSceneTimeline,
        "favorite", SendMessageToWX.Req.WXSceneFavorite
      )
    );
  }

  /**
   * 向微信注册APP
   *
   * @param appId   微信app id
   * @param promise js 异步回调
   */
  @ReactMethod
  public void registerApp(String appId, Promise promise) {
    weChatApi = WXAPIFactory.createWXAPI(getReactApplicationContext(), appId, true);
    if (null != getCurrentActivity()) {
      weChatApi.handleIntent(getCurrentActivity().getIntent(), this);
    } else {
      Log.w(TAG, "current activity is null.");
    }
    boolean result = weChatApi.registerApp(appId);
    Log.d(TAG, "向微信注册APP, result = " + result);
  }

  @ReactMethod
  public void share(ReadableMap options, Promise promise) {
    if (!options.hasKey("scene")) {
      Log.e(TAG, "微信分享失败,不正确的分享场景");
      promise.reject("3000", "微信分享失败,不正确的分享场景");
      return;
    }

    if (!options.hasKey("type")) {
      Log.e(TAG, "微信分享失败,不正确的类型");
      promise.reject("3001", "不正确的消息类型");
      return;
    }

    ShareType type = ShareType.values()[options.getInt("type")];
    if (null == type) {
      Log.e(TAG, "微信分享失败,不正确的类型");
      promise.reject("3001", "不正确的消息类型");
      return;
    }

    WXMediaMessage message = new WXMediaMessage();
    if (options.hasKey("title")) {
      message.title = options.getString("title");
    }
    if (options.hasKey("description")) {
      message.description = options.getString("description");
    }
    if (options.hasKey("mediaTagName")) {
      message.mediaTagName = options.getString("mediaTagName");
    }
    if (options.hasKey("messageAction")) {
      message.messageAction = options.getString("messageAction");
    }
    if (options.hasKey("messageExt")) {
      message.messageExt = options.getString("messageExt");
    }

    if (options.hasKey("image")) {
      try {
        Bitmap bitmap = BitmapFactory.decodeStream(new URL(options.getString("image")).openStream());
        message.setThumbImage(bitmap);
      } catch (IOException e) {
        Log.e(TAG, "设置图片失败", e);
      }
    }

    switch (type) {
      case TEXT:
        WXTextObject textObject = new WXTextObject();
        textObject.text = options.getString("text");
        message.mediaObject = textObject;
        break;
      case IMAGE:
        WXImageObject imageObject = new WXImageObject();
        imageObject.imageData = message.thumbData;
        message.mediaObject = imageObject;
        break;
      case MUSIC:
        WXMusicObject musicObject = new WXMusicObject();
        musicObject.musicUrl = "";
        message.mediaObject = musicObject;
        break;
      case VIDEO:
        break;
      case WEBPAGE: {
        WXWebpageObject webpageObject = new WXWebpageObject();
        if (!options.hasKey("webPageUrl")) {
          Log.e(TAG, "微信分享失败,缺少webPageUrl");
          promise.reject("3001", "缺少webPageUrl");
          return;
        }
        webpageObject.webpageUrl = options.getString("webPageUrl");
        message.mediaObject = webpageObject;
        break;
      }
    }

    SendMessageToWX.Req req = new SendMessageToWX.Req();
    req.scene = options.getInt("scene");
    req.message = message;
    req.transaction = UUID.randomUUID().toString();
    boolean success = weChatApi.sendReq(req);
    promise.resolve(success);
  }

  @Override
  public void onReq(BaseReq baseReq) {
    Log.d(TAG, "onReq transaction -> " + baseReq.transaction);
  }

  @Override
  public void onResp(BaseResp baseResp) {
    Log.d(TAG, "onResp transaction -> " + baseResp.transaction);
    WritableMap map = Arguments.createMap();
    map.putInt("errCode", baseResp.errCode);
    map.putString("errStr", baseResp.errStr);
    map.putString("openId", baseResp.openId);
    map.putString("transaction", baseResp.transaction);

    if (baseResp instanceof SendAuth.Resp) {
      SendAuth.Resp resp = (SendAuth.Resp) (baseResp);

      map.putString("type", "SendAuth.Resp");
      map.putString("code", resp.code);
      map.putString("state", resp.state);
      map.putString("url", resp.url);
      map.putString("lang", resp.lang);
      map.putString("country", resp.country);
    } else if (baseResp instanceof SendMessageToWX.Resp) {
      SendMessageToWX.Resp resp = (SendMessageToWX.Resp) (baseResp);
      map.putString("type", "SendMessageToWX.Resp");
    } else if (baseResp instanceof PayResp) {
      PayResp resp = (PayResp) (baseResp);
      map.putString("type", "PayReq.Resp");
      map.putString("returnKey", resp.returnKey);
    }

    sendEvent("WeChatResp", map);
  }

  public enum ShareType {
    TEXT,
    IMAGE,
    MUSIC,
    VIDEO,
    WEBPAGE
  }

}

File: ios/SitbWeChat/WeChatManager.m
//
//  WeChatManager.m
//  WeChatManager
//
//  Created by 田尘殇 on 2017/5/17.
//  Copyright © 2017年 Sitb Software. All rights reserved.
//

#import "WeChatManager.h"
#import <React/RCTLog.h>
#import <React/RCTImageLoader.h>

typedef NS_ENUM(NSInteger, ShareType) {
    TEXT = 0,
    IMAGE = 1,
    MUSIC = 2,
    VIDEO = 3,
    WEBPAGE = 4
};

@implementation WeChatManager

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE(WeChatManager);

- (NSDictionary<NSString *, id> *)constantsToExport {
    return @{
            @"type": @{
                    @"text": @(TEXT),
                    @"image": @(IMAGE),
                    @"music": @(MUSIC),
                    @"video": @(VIDEO),
                    @"webPage": @(WEBPAGE)
            },
            @"scene": @{
                    @"session": @(WXSceneSession),
                    @"timeLine": @(WXSceneTimeline),
                    @"favorite": @(WXSceneFavorite)
            }
    };
}

+ (BOOL)requiresMainQueueSetup {
    return NO;
}

RCT_EXPORT_METHOD(
            registerApp:
            (NSString *) appId) {
    BOOL result = [WXApi registerApp:appId];
    RCTLog(@"向微信注册APP %@, result -> %d", appId, result);
}


/**
 * 分享给指定的回话
 * @param resolve 异步执行成功
 * @return
 */
RCT_EXPORT_METHOD(
            share:
            (NSDictionary *) data
            resolve:
            (RCTPromiseResolveBlock) resolve
            reject:
            (RCTPromiseRejectBlock) reject) {
    [self shareToWeChatWithData:data scene:((NSNumber *) data[@"scene"]).intValue resolve:resolve reject:reject];
}

- (void)shareToWeChatWithData:(NSDictionary *)data
                        scene:(int)scene
                      resolve:(RCTPromiseResolveBlock)resolve
                       reject:(RCTPromiseRejectBlock)reject {

    NSString *image = data[@"image"];
    if (image.length && _bridge.imageLoader) {
        CGSize size;
        CGFloat thumbImageSize = 80;
        if (data[@"thumbImageSize"]) {
            thumbImageSize = [data[@"thumbImageSize"] floatValue];
        }
        size = CGSizeMake(thumbImageSize, thumbImageSize);

        [_bridge.imageLoader loadImageWithURLRequest:[RCTConvert NSURLRequest:image]
                                                size:size
                                               scale:1
                                             clipped:FALSE
                                          resizeMode:(RCTResizeMode) UIViewContentModeScaleToFill
                                       progressBlock:nil
                                    partialLoadBlock:nil
                                     completionBlock:^(NSError *error, UIImage *img) {
                                         [self shareToWeChatWithData:data scene:scene image:img resolve:resolve reject:reject];
                                     }
        ];
    } else {
        [self shareToWeChatWithData:data scene:scene image:nil resolve:resolve reject:reject];
    }


}

- (void)shareToWeChatWithData:(NSDictionary *)data
                        scene:(int)scene
                        image:(UIImage *)image
                      resolve:(RCTPromiseResolveBlock)resolve
                       reject:(RCTPromiseRejectBlock)reject {
    SendMessageToWXReq *req = [SendMessageToWXReq new];
    req.scene = scene;


    ShareType type = (ShareType) ((NSNumber *) data[@"type"]).intValue;

    if (type == TEXT) {
        NSString *text = data[@"text"];
        req.bText = YES;
        req.text = text;
    } else {
        req.bText = NO;

        WXMediaMessage *message = [WXMediaMessage message];
        message.title = data[@"title"];
        message.description = data[@"description"];
        message.mediaTagName = data[@"mediaTagName"];
        message.messageAction = data[@"messageAction"];
        message.messageExt = data[@"messageExt"];
        if (image) {
            [message setThumbImage:image];
        }

        switch (type) {
            case TEXT:
                break;
            case IMAGE: {
                WXImageObject *imageObject = [WXImageObject object];
                imageObject.imageData = UIImagePNGRepresentation(image);
                message.mediaObject = imageObject;
                break;
            }
            case MUSIC: {
                break;
            }
            case VIDEO:
                break;
            case WEBPAGE: {
                WXWebpageObject *webpageObject = [WXWebpageObject object];
                webpageObject.webpageUrl = data[@"webPageUrl"];
                message.mediaObject = webpageObject;
                break;
            }
        }
        req.message = message;
    }
    BOOL success = [WXApi sendReq:req];

    if (success) {
        RCTLog(@"微信分享成功");
        resolve(@"");
    } else {
        RCTLog(@"微信分享失败");
        reject(@"3000", @"微信分享失败", nil);
    }
}


- (void)onReq:(BaseReq *)req {
    if ([req isKindOfClass:[GetMessageFromWXReq class]]) {
        // 微信请求App提供内容， 需要app提供内容后使用sendRsp返回
        NSString *strTitle = [NSString stringWithFormat:@"微信请求App提供内容"];
        NSString *strMsg = @"微信请求App提供内容，App要调用sendResp:GetMessageFromWXResp返回给微信";
        RCTLog(@"%@,%@", strTitle, strMsg);
    } else if ([req isKindOfClass:[ShowMessageFromWXReq class]]) {
        ShowMessageFromWXReq *temp = (ShowMessageFromWXReq *) req;
        WXMediaMessage *msg = temp.message;

        //显示微信传过来的内容
        WXAppExtendObject *obj = msg.mediaObject;

        NSString *strTitle = [NSString stringWithFormat:@"微信请求App显示内容"];
        NSString *strMsg = [NSString stringWithFormat:@"标题：%@ \n内容：%@ \n附带信息：%@ \n缩略图:%u bytes\n\n", msg.title, msg.description, obj.extInfo, msg.thumbData.length];
        RCTLog(@"%@,%@", strTitle, strMsg);
    } else if ([req isKindOfClass:[LaunchFromWXReq class]]) {
        //从微信启动App
        NSString *strTitle = [NSString stringWithFormat:@"从微信启动"];
        NSString *strMsg = @"这是从微信启动的消息";

        RCTLog(@"%@,%@", strTitle, strMsg);
    }
}

- (void)onResp:(BaseResp *)resp {
    if ([resp isKindOfClass:[SendMessageToWXResp class]]) {
        NSString *strTitle = [NSString stringWithFormat:@"发送媒体消息结果"];
        NSString *strMsg = [NSString stringWithFormat:@"errcode:%d", resp.errCode];
        RCTLog(@"%@,%@", strTitle, strMsg);
    }
}

@end
